botocore/utils.py

```diff
@@ -260,70 +265,108 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
             proxies=get_environ_proxies(self._url),
         )
 
-    def _get_request(self, url, timeout, num_attempts=1):
+    def retrieve_iam_role_credentials(self):
+        try:
+            role_name = self._get_iam_role()
+            credentials = self._get_credentials(role_name)
+            if self._contains_all_credential_fields(credentials):
```

Comment:
Why isn't `_contains_all_credential_fields` not part of the `self._does_not_contain_credentials` retry check?  It seems like part of checking if the response does not contain credentials is to also check that all the credential keys are present in the response.

Reply:
I put it in the [comment](https://github.com/boto/botocore/pull/1569/files/f5611ae5bfe0b374d3ce9ba85a67cb9590fcdcb1#diff-6c5fe2c0e8325d002e7c164f10a30dbcR281). I don't think we should be retrying these cases. I think I'm going to rename these methods to make it a bit clearer. I can see how the names make it confusing on why the `_contains_all_credential_fields` is not in the `_does_not_contain_credentials` method